jgrapht-core/src/main/java/org/jgrapht/alg/planar/ElementList.java

```diff
@@ -0,0 +1,438 @@
+     *
+     * @param list the list to prepend
+     */
+    public void prepend(ElementList<E> list) {
+        if (!list.isEmpty()) {
+            if (!isEmpty()) {
+                linkBefore(head, list);
+            }
+            head = list.head;
+            size += list.size;
+
+            list.size = 0;
+            list.head = null;
+        }
+    }
+
+    /**
+     * Appends the {@code value} to the end of this list. Returns the list node allocated for storing
+     * the {@code value}
+     *
+     * @param value the value to add
+     * @return the list node allocated for storing the {@code value}
+     */
+    public ListNode<E> append(E value) {
+        ListNode<E> result = new ListNode<>(value);
+        if (isEmpty()) {
+            insertFirst(result);
+        } else {
+            linkBefore(head, result);
+        }
+        ++size;
+        return result;
+    }
+
+    /**
+     * Prepends the {@code value} to the beginning of this list. Returns the list node allocated for storing
+     * the {@code value}. The returned value is the new head of the list.
+     *
+     * @param value the value to add
+     * @return the list node allocated for storing the {@code value}
+     */
+    public ListNode<E> prepend(E value) {
+        ListNode<E> result = append(value);
+        head = result;
+        return result;
+    }
+
+    /**
+     * Removes the {@code node} from this list. The running time of this method is $\mathcal{O}(1)$.
+     * The behaviour of this method is undefined if the {@code node} doesn't belong to the list.
+     *
+     * @param node the node to remove from this list
+     */
+    public void remove(ListNode<E> node) {
+        if (size == 1) {
+            head = null;
+        } else {
+            if (node == head) {
+                head = node.next;
+            }
+        }
+        unlink(node);
+        --size;
+    }
+
+    /**
+     * Returns the first element in this list. If this list is empty, throws {@link NoSuchElementException}.
+     *
+     * @return the first element in this list.
+     */
+    public E getFirstElement() {
+        if (isEmpty()) {
+            throw new NoSuchElementException();
+        }
+        return head.value;
+    }
+
+    /**
+     * Returns the first element in this list. If this list is empty, throws {@link NoSuchElementException}
+     *
+     * @return the first element in this list.
+     */
+    public E getLastElement() {
+        if (isEmpty()) {
+            throw new NoSuchElementException();
+        }
+        return head.prev.value;
+    }
+
+    /**
+     * Finds and returns the list node allocate for the {@code element}. If this list doesn't contain
+     * the {@code element}, returns {@code null}
+     *
+     * @param element the element of this list to search the list node of.
+     * @return the list node allocate for the {@code element}, or {@code null} if this list doesn't contain
+     * the {@code element}
+     */
```

Comment:
Mention complexity of this method and the fact that equality checks uses method `equals()`.